File: app/src/main/java/com/soundcloud/android/ads/AdRequestData.java
package com.soundcloud.android.ads;

import com.google.auto.value.AutoValue;
import com.soundcloud.android.model.Urn;
import com.soundcloud.java.optional.Optional;

import java.util.UUID;

@AutoValue
public abstract class AdRequestData {
    private static AdRequestData create(Optional<Urn> monetizableTrackUrn, Optional<String> kruxSegments)  {
        return new AutoValue_AdRequestData(UUID.randomUUID().toString(), monetizableTrackUrn, kruxSegments);
    }

    public static AdRequestData forPlayerAd(Urn monetizableTrackUrn, Optional<String> kruxSegments) {
        return create(Optional.of(monetizableTrackUrn), kruxSegments);
    }

    public static AdRequestData forPageAds(Optional<String> kruxSegments) {
        return create(Optional.absent(), kruxSegments);
    }

    public abstract String getRequestId();

    public abstract Optional<Urn> getMonetizableTrackUrn();

    public abstract Optional<String> getKruxSegments();
}

File: app/src/main/java/com/soundcloud/android/ads/AdRequestData.kt
package com.soundcloud.android.ads

import com.soundcloud.android.model.Urn
import com.soundcloud.java.optional.Optional
import java.util.*

data class AdRequestData(
        val monetizableTrackUrn: Urn? = null,
        val kruxSegments: String? = null,
        val requestId: String = UUID.randomUUID().toString()
) {
    companion object {
        @JvmStatic
        fun forPlayerAd(monetizableTrackUrn: Urn, kruxSegments: Optional<String> = Optional.absent()): AdRequestData {
            return AdRequestData(monetizableTrackUrn, kruxSegments.orNull())
        }

        @JvmStatic
        fun forPageAds(kruxSegments: Optional<String> = Optional.absent()): AdRequestData {
            return AdRequestData(kruxSegments = kruxSegments.orNull())
        }
    }
}
